rules/no-assign-parameter.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * @fileoverview Do not assign to parameters.
+ * @author Simon Hajjar <simon.j.hajjar@gmail.com>
+ */
+
+'use strict';
+
+module.exports = {
+
+    meta: {
+
+        docs: {
+            recommended: true,
+            type: 'error',
+            description: 'Do not assign to parameters'
+        },
+
+        schema: []
+    },
+
+    create: function (context) {
+
+        function inspectFunctionDeclaration (emitted) {
+            if (emitted.exit) { return; }
+
+            var node = emitted.node;
+            let params = node.params.map(x => x['id']);
+            for (let statement of node.body.body) {
+                if(statement['expression']['type'] === 'AssignmentExpression' &&
```

Comment:
If I understand correctly, you traverse through all the top-level stmts inside the function block. Couple of things:
- Every statement is not guaranteed to have the `expression` attribute. You should first ensure that the statement you're checking is an `ExpressionStatement`. Otherwise results in exception
- This technique only catches top-level assignments. If the assignment is happening inside a for loop or if/else block, I don't think it will catch it.

Please correct me if I'm wrong above^